src/hotspot/share/gc/shared/collectedHeap.hpp

```diff
@@ -518,6 +521,20 @@ class CollectedHeap : public CHeapObj<mtGC> {
   virtual HeapWord* allocate_loaded_archive_space(size_t size) { return nullptr; }
   virtual void complete_loaded_archive_space(MemRegion archive_space) { }
 
+  // Commit the heap memory for CDS heap archive area according to the requested size.
+  // Preferred address is treated as a hint for the location of the archive area in the heap.
+  // The returned memory range may or may start at the preferred address.
+  // Return MemRegion corresponding to the commited heap memory.
+  virtual HeapWord* alloc_archive_space(size_t word_size, HeapWord* preferred_addr) { return nullptr; }
+
+  // This must be called after alloc_archive_space, and after class loading has occurred.
+  // GC policy can take necessary actions to make the archive space usable,
+  // such as inserting filler objects to make it parseable, or updating block offset table.
+  virtual void fixup_archive_space(MemRegion range) { return; }
```

Comment:
Small nit about the naming of these callbacks. "fixup_archive_space" implies knowledge about what the collector does, as does "handle_archive_space_failure". But you want to abstract away collector work from CDS. @iklam already remarked on "handle..". 

I see these methods both more as callbacks, and as such they could be named after the point at which they are called, e.g. "on_cds_post_load" and "on_cds_load_error". Or similar. Named after when they are called, but assuming no knowledge of what the collector does at that point, if anything at all.

Up to you. This is very bikesheddy, I'm also fine with the names as they are.